File: internal/service/scheduler_service.go
package service
import (
	"log"
	"time"
	"autosmm/internal/models"
	"autosmm/internal/repository"
	"github.com/google/uuid"
)
type SchedulerService struct { postRepo *repository.PostRepository }
func NewSchedulerService(repo *repository.PostRepository) *SchedulerService { return &SchedulerService{postRepo: repo} }
func (s *SchedulerService) SchedulePost(userID uuid.UUID, req models.PostRequest) (*models.Post, error) {
	post := &models.Post{
		ID: uuid.New(), UserID: userID, Content: req.Content,
		ScheduledAt: req.ScheduledAt, Status: "scheduled",
	}
	err := s.postRepo.CreatePost(post, req.TargetAccountIDs)
	if err != nil { return nil, err }
	return post, nil
}
func (s *SchedulerService) GetUserPosts(userID uuid.UUID) ([]models.Post, error) {
	return s.postRepo.GetUserPosts(userID)
}
func (s *SchedulerService) StartWorker() {
	ticker := time.NewTicker(30 * time.Second)
	go func() {
		for range ticker.C { s.publishDuePosts() }
	}()
}
func (s *SchedulerService) publishDuePosts() {
	posts, err := s.postRepo.GetPendingPosts()
	if err != nil { log.Printf("Error fetching pending posts: %v", err); return }
	for _, post := range posts {
		log.Printf("🚀 Publishing post %s: %s", post.ID, post.Content)
		if err := s.postRepo.MarkPostPublished(post.ID); err != nil {
			log.Printf("Error marking post %s as published: %v", post.ID, err)
		}
	}
}

File: src/services/schedulerService.js

const db = require('../db');
const cron = require('node-cron');
class SchedulerService {
  async schedulePost(userId, { content, scheduledAt, targetAccountIds }) {
    const postQuery = 'INSERT INTO posts (user_id, content, scheduled_at, status) VALUES ($1, $2, $3, \'scheduled\') RETURNING id';
    const postResult = await db.query(postQuery, [userId, content, scheduledAt]);
    const postId = postResult.rows[0].id;
    for (const accId of targetAccountIds) {
      await db.query('INSERT INTO post_targets (post_id, account_id, status) VALUES ($1, $2, \'pending\')', [postId, accId]);
    }
    return { id: postId, status: 'scheduled' };
  }
  async getUserPosts(userId) {
    const result = await db.query('SELECT * FROM posts WHERE user_id = $1 ORDER BY scheduled_at ASC', [userId]);
    return result.rows;
  }
  startWorker() {
    cron.schedule('*/30 * * * * *', async () => {
      try {
        const result = await db.query("SELECT * FROM posts WHERE status = 'scheduled' AND scheduled_at <= NOW()");
        for (const post of result.rows) {
          await db.query('UPDATE posts SET status = \'published\', published_at = NOW() WHERE id = $1', [post.id]);
        }
      } catch (err) { console.error('Worker error:', err); }
    });
  }
}
module.exports = new SchedulerService();
